src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PENamespaceSymbol.cs

```diff
@@ -89,29 +91,42 @@ private ImmutableArray<NamedTypeSymbol> GetMemberTypesPrivate()
 
         public sealed override ImmutableArray<Symbol> GetMembers(ReadOnlyMemory<char> name)
         {
+            var (lastName, lastResult) = _lastGetMembersResult;
```

Comment:
This is based on @sharwell suggestion in https://github.com/dotnet/roslyn/issues/68946

Reply:
this has no threadsafety, so it could lead to torn read/writes between different callers asking for GetMembers with different names.  Very scary :)

Reply:
Should I just surround this with a lock?

Reply:
i would be VERY wary about adding a lock into this hot path.  It would make me super uncomfortable.

What you can consider doing *instead* is not using a value-tuple, but genuinely just use a normal System.Tuple (or just a newly defined simple class) for this purpose. 

That said, i'm rusty on where .Net currently lands on threading, and if it allows write reordering to be visible across threads.  If so, then even a reference-type here might not be ok as you might see the write to the field, before seeing all the writes within the value written into the field.  Would need someone more confident to specify how to best handle that (for exmaple, if Interlocked operations shoudl be used, or if the field shoudl be `volatile`).

Reply:
Maybe @stephentoub could advise here?

Reply:
I wouldn't add a lock, and we also want to avoid a new allocation. We can avoid the need to write two fields by omitting the `lastName` from the cache; this name is just the `Name` of the first element of `lastResult`.